libcamera/src/control.rs

```diff
@@ -456,13 +458,88 @@ impl<'a> Drop for ControlInfoMapIter<'a> {
     }
 }
 
+#[derive(Debug, Clone, Copy, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum ControlDirection {
+    /// Input flag (1<<0)
+    In = LIBCAMERA_CONTROL_DIRECTION_IN,
+    /// Output flag (1<<1)
+    Out = LIBCAMERA_CONTROL_DIRECTION_OUT,
+    /// Input and output flags combined (1<<0 | 1<<1)
+    InOut = LIBCAMERA_CONTROL_DIRECTION_IN | LIBCAMERA_CONTROL_DIRECTION_OUT,
+}
+
 impl ControlId {
     pub fn name(&self) -> String {
         unsafe { CStr::from_ptr(libcamera_control_name_from_id(self.id())) }
             .to_str()
             .unwrap()
             .into()
     }
+
+    fn as_ptr(&self) -> *mut libcamera_control_id_t {
+        unsafe { libcamera_control_from_id(self.id()) as *mut _ }
+    }
+
+    pub fn vendor(&self) -> String {
+        unsafe {
+            let ctrl = self.as_ptr();
+            if ctrl.is_null() {
+                String::new()
+            } else {
+                let ptr = libcamera_control_id_vendor(ctrl);
+                if ptr.is_null() {
+                    String::new()
+                } else {
+                    CStr::from_ptr(ptr).to_string_lossy().to_string()
+                }
+            }
+        }
+    }
+
+    pub fn control_type(&self) -> ControlType {
+        let raw = unsafe { libcamera_control_id_type(self.as_ptr()) } as u32;
+        ControlType::try_from(raw).expect("Unknown ControlType")
+    }
+
+    pub fn direction(&self) -> ControlDirection {
+        let raw = unsafe { libcamera_control_id_direction(self.as_ptr()) } as u32;
+        ControlDirection::try_from(raw).expect("Unknown libcamera_control_direction value")
+    }
+
+    pub fn is_input(&self) -> bool {
+        unsafe { libcamera_control_id_is_input(self.as_ptr()) }
+    }
+
+    pub fn is_output(&self) -> bool {
+        unsafe { libcamera_control_id_is_output(self.as_ptr()) }
+    }
+
+    pub fn is_array(&self) -> bool {
+        unsafe { libcamera_control_id_is_array(self.as_ptr()) }
+    }
+
+    pub fn size(&self) -> usize {
+        unsafe { libcamera_control_id_size(self.as_ptr()) }
+    }
+
+    pub fn enumerators_map(&self) -> HashMap<i32, String> {
```

Comment:
I think this would work better as an iterator, see [ControlListRefIterator](https://github.com/lit-robotics/libcamera-rs/blob/main/libcamera/src/control.rs#L378) implementation to iterate a C++ map. Current implementation will cost `O(2n log n)` to collect everything into hashmap, because `std::advance(it, index)` is `O(log n)` for C++ map, which is called for each entry twice. I know that this is not a performance critical path, but I would like it to be consistent inside the repo. A separate convenience function can just `.collect()` rust iterator into a `HashMap<i32, String>`.